[14.0][ADD] web_tree_customized_field_list

ilyasProgrammer opened the pull request.

ilyasProgrammer committed:
[ADD] web_tree_customized_field_list

elvise commented:
@ilyasProgrammer can you rebase ?

elvise commented:
/ocabot rebase

ilyasProgrammer commented:
@elvise done

elvise commented:
@ilyasProgrammer please rebase :)

francesco-ooops reviewed (approved):
Functional ok!

Often a customer asks to add a field to tree view, and this requires dev time. This module allows functionals to be independent on adding fields to tree view.

geomer198 commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,139 @@
+
+    name = fields.Char(required=True)
+    model_id = fields.Many2one("ir.model", required=True, ondelete="cascade")
+    model_name = fields.Char(compute="_compute_model_name", store=True)
```
For what this field have compute attribute? Could you make this field with related attribute?
```py
model_name = fields.Char(related="model_id.model")
```
I think this solution is easier.

geomer198 commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,139 @@
+    @api.constrains("model_id")
+    def _check_unique_model(self):
+        for rec in self:
+            same_model_recs = self.env["custom.list.view"].search(
```
I think better use `search_count` method. Method `search` with `count=True` argument return clean data, without running `browse` method.

geomer198 commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,139 @@
+    @api.model_create_multi
+    def create(self, vals_list):
+        res = super().create(vals_list)
+        res.original_arch = res.list_view_id.arch
```
Please rename `res` variable, because `create` method return record or record set. Better name will be `records` here.
I think is wrong, `res` variable is record set then you need make like this:
```py
for record in records:
    record.original_arch = record.list_view_id.arch
```

geomer198 commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,139 @@
+        res.original_arch = res.list_view_id.arch
+        return res
+
+    def write(self, vals):
```
For what this method here? Please remove unused code.

geomer198 commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,139 @@
+        self.ensure_one()
+        doc = etree.XML(self.original_arch)
+        for mod_line in self.line_ids:
+            target = mod_line.before.name if mod_line.before else mod_line.after.name
```
This line is correct but will make better:
```py
mod_line.before and mod_line.before.name or mod_line_after.name
```

geomer198 commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,139 @@
+        self.list_view_id.arch = new_arch
+
+
+class CustomListViewLine(models.Model):
```
I think better this class get out of separate file with name `customer_list_view_line.py`.

geomer198 replied:
> I think better this class get out of separate file with name `customer_list_view_line.py`.

@ilyasProgrammer What do you think about it?

geomer198 commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,139 @@
+    after = fields.Many2one("ir.model.fields", "Place After")
+    before = fields.Many2one("ir.model.fields", "Place Before")
+    optional = fields.Selection(
+        [("show", "show"), ("hide", "hide"), ("", "")], default=""
```
Why do you added empty option?
If you remove `default` argument and empty option then by default in selection empty option is auto added.

geomer198 commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,139 @@
+    label = fields.Char()
+    widget = fields.Selection(
+        [
+            ("many2many_tags", "many2many_tags"),
```
I think use technical name for name of selection element  is not good idea.

geomer198 replied:
This comment is not resolved.

geomer198 commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,139 @@
+            rec.fields_domain = json.dumps([("id", "in", field_ids.ids)])
+
+
+class IrUiView(models.Model):
```
I think better this class get out of separate file with name `ir_ui_view.py`.

geomer198 commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,139 @@
+    _inherit = "ir.ui.view"
+
+    def name_get(self):
+        res = []
```
It's correct but you can do in one line:
```py
return [(rec.id, rec.xml_id) for rec in self]
```

geomer198 commented on customized_list_view/tests/test_customized_list_view.py:
```diff
@@ -0,0 +1,68 @@
+        cls.users_model = cls.env["ir.model"]._get("res.users")
+        cls.view_users_tree = cls.env.ref("base.view_users_tree")
+
+    def test_all_customized_list_view(self):
```
Could you add docstring where described what you do in this test. 
Please add message for each assert case.

geomer198 replied:
This comment is not resolved.

geomer198 reviewed (changes requested):
Could you please fix code by my comments.

ilyasProgrammer commented:
@geomer198 Thanks for the review. Updated according to your suggestions.

HekkiMelody commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,120 @@
+                        "Only one record per model is allowed. "
+                        "Please modify existing record instead."
+                    )
+                )
```
**suggestion (non-blocking)**: I think you could implement a simpler SQL `unique` constraint here.

**question**: Shouldn't the uniqueness constraint apply to `list_view_id` rather than `model_id`?

ilyasProgrammer replied:
Agree

HekkiMelody commented on customized_list_view/security/ir.model.access.csv:
```diff
@@ -0,0 +1,5 @@
+clw1,clw1,model_custom_list_view,base.group_system,1,1,1,1
+clw2,clw2,model_custom_list_view,base.group_user,1,0,0,0
+clwl1,clwl1,model_custom_list_view_line,base.group_system,1,1,1,1
+clwl2,clwl2,model_custom_list_view_line,base.group_user,1,0,0,0
```
**question**: does `base.group_user` actually need access to this?

ilyasProgrammer replied:
No. Removed.

HekkiMelody reviewed (commented):
Code review, only a few minor remarks.

However, I should point out that the customizations do not survive an upgrade. In other words, when any module is upgraded, the views of that module are reverted back to what is written in the XML file, and it would be necessary to press "Apply Changes" again to make them appear again (Tested on runboat)

ilyasProgrammer commented:
> Code review, only a few minor remarks.
> 
> However, I should point out that the customizations do not survive an upgrade. In other words, when any module is upgraded, the views of that module are reverted back to what is written in the XML file, and it would be necessary to press "Apply Changes" again to make them appear again (Tested on runboat)

Thats right. Updated description and added roadmap.

HekkiMelody commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,112 @@
+    original_arch = fields.Text(readonly=True)
+
+    _sql_constraints = [
+        (
+            "unique_model_list_view_rec",
+            "UNIQUE(model_id, list_view_id)",
+            "Only one record per model is allowed. "
+            "Please modify existing record instead.",
+        )
```
This is really "Only one record per view is allowed", and the uniqueness constraint could be applied to `list_view_id` only, IMO.

HekkiMelody reviewed (approved):
Approved with the above caveat that it may be very counterintuitive to use.

HekkiMelody commented on customized_list_view/models/ir_module.py:
```diff
@@ -0,0 +1,18 @@
+class Module(models.Model):
+    _inherit = "ir.module.module"
+
+    def _button_immediate_function(self, function):
```
**praise** this is an elegant way of making the changes persistent

**thought**: not all changes may be applied. at any given time. I'd say we could either add the `active` field to the `custom.list.view` model, or another boolean field to keep track of which changes need to be applied and which ones should not.

HekkiMelody commented:
Code review LGTM

geomer198 commented on customized_list_view/models/ir_module.py:
```diff
@@ -0,0 +1,18 @@
+    def _button_immediate_function(self, function):
+        res = super(Module, self)._button_immediate_function(function)
+        views_mods = self.env["custom.list.view"].search([])
+        for vm in views_mods:
```
This solution is not optimal.
Please change this implementation without 2 cycles. If you want remove all 'custom.list.view.line' records without `field_id` value, you can use search for this. I think it's faster then you solution.

geomer198 reviewed (changes requested):
Could you check my review?

geomer198 commented on customized_list_view/models/custom_list.py:
```diff
@@ -0,0 +1,64 @@
+    active = fields.Boolean(default=True)
+    name = fields.Char(required=True)
+    model_id = fields.Many2one("ir.model", required=True, ondelete="cascade")
+    model_name = fields.Char(related="model_id.model", store=True)
```
For what you use `store` here?

ilyasProgrammer replied:
To avoid recompute.

geomer198 commented on customized_list_view/tests/test_customized_list_view.py:
```diff
@@ -0,0 +1,83 @@
+        cls.users_model = cls.env["ir.model"]._get("res.users")
+        cls.view_users_tree = cls.env.ref("base.view_users_tree")
+
+    def test_all_customized_list_view(self):
```
Could you add doc string for this test method, which describe what this method do?

ilyasProgrammer replied:
The method is for tests. There is comment for each subtest and message for each assert. It is sufficient.

geomer198 reviewed (changes requested):
Please, add little fix.

geomer198 reviewed (approved):
LGTM!

francesco-ooops commented:
@pedrobaeza could you merge this one? thanks!

pedrobaeza commented:
Being a modification of the web client, it should be prefixed with `web_` IMO. A name can be `web_tree_customized_field_list`.

francesco-ooops commented:
@pedrobaeza should it be in this repo or in OCA/web in your opinion?

pedrobaeza commented:
Good question... Let's keep it here as is less crowded.

francesco-ooops reviewed (approved):
LGTM

@pedrobaeza shall we go?

pedrobaeza commented:
/ocabot merge nobump
